File: app/src/main/java/de/karzek/diettracker/presentation/main/cookbook/recipeManipulation/automatedIngredientSearch/AutomatedIngredientSearchActivity.java
package de.karzek.diettracker.presentation.main.cookbook.recipeManipulation.automatedIngredientSearch;

import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.HashMap;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import de.karzek.diettracker.R;
import de.karzek.diettracker.data.cache.model.GroceryEntity;
import de.karzek.diettracker.presentation.TrackerApplication;
import de.karzek.diettracker.presentation.common.BaseActivity;
import de.karzek.diettracker.presentation.main.cookbook.recipeManipulation.automatedIngredientSearch.adapter.IngredientSearchListAdapter;
import de.karzek.diettracker.presentation.main.cookbook.recipeManipulation.automatedIngredientSearch.adapter.itemWrapper.IngredientSearchItemWrapper;
import de.karzek.diettracker.presentation.model.IngredientDisplayModel;
import de.karzek.diettracker.presentation.model.ManualIngredientDisplayModel;
import de.karzek.diettracker.presentation.model.RecipeDisplayModel;
import de.karzek.diettracker.presentation.search.grocery.GrocerySearchActivity;
import de.karzek.diettracker.presentation.search.grocery.barcodeScanner.BarcodeScannerActivity;
import de.karzek.diettracker.presentation.util.Constants;

import static de.karzek.diettracker.presentation.util.Constants.INVALID_ENTITY_ID;

public class AutomatedIngredientSearchActivity extends BaseActivity implements AutomatedIngredientSearchContract.View {

    public static final String EXTRA_INDEX = "EXTRA_INDEX";
    public static final String EXTRA_AMOUNT = "EXTRA_AMOUNT";
    public static final String EXTRA_UNIT_ID = "EXTRA_UNIT_ID";
    public static final String EXTRA_GROCERY_ID = "EXTRA_GROCERY_ID";
    public static final String EXTRA_RECIPE = "EXTRA_RECIPE";

    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.loading_view)
    FrameLayout loadingView;
    @BindView(R.id.recycler_view)
    RecyclerView recyclerView;
    @BindView(R.id.save_button)
    Button saveButton;

    @Inject
    AutomatedIngredientSearchContract.Presenter presenter;

    public static Intent newIntent(Context context, RecipeDisplayModel recipe) {
        Intent intent = new Intent(context, AutomatedIngredientSearchActivity.class);
        intent.putExtra(EXTRA_RECIPE, recipe);

        return intent;
    }

    @Override
    protected void setupActivityComponents() {
        TrackerApplication.get(this).createAutomatedIngredientSearchComponent().inject(this);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_automated_ingredient_search);
        ButterKnife.bind(this);

        setupSupportActionBar();
        setupRecyclerView();
        disableSaveButton();

        presenter.setView(this);
        RecipeDisplayModel recipeDisplayModel = getIntent().getExtras().getParcelable(EXTRA_RECIPE);
        presenter.setRecipe(recipeDisplayModel);
        if(recipeDisplayModel.getId() != INVALID_ENTITY_ID)
            presenter.startEdit();
        else
            presenter.start();
    }

    private void disableSaveButton() {
        saveButton.setAlpha(.5f);
        saveButton.setClickable(false);
    }

    @Override
    public void enableSaveButton() {
        saveButton.setAlpha(1.f);
        saveButton.setClickable(true);
    }

    private void setupSupportActionBar() {
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
        getSupportActionBar().setHomeAsUpIndicator(getResources().getDrawable(R.drawable.ic_back_arrow_white, null));
        getSupportActionBar().setTitle(getString(R.string.automated_ingredient_search_title));
    }

    private void setupRecyclerView() {
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(new IngredientSearchListAdapter(presenter, presenter, presenter));
    }

    @Override
    public void setupViewsInRecyclerView(ArrayList<IngredientDisplayModel> ingredients) {
        ArrayList<IngredientSearchItemWrapper> views = new ArrayList<>();

        for (IngredientDisplayModel ingredient : ingredients) {
            if (ingredient instanceof ManualIngredientDisplayModel) {
                views.add(new IngredientSearchItemWrapper(IngredientSearchItemWrapper.ItemType.LOADING_INGREDIENT, ingredient));
            } else {
                views.add(new IngredientSearchItemWrapper(IngredientSearchItemWrapper.ItemType.FOUND_INGREDIENT, ingredient));
            }
        }

        ((IngredientSearchListAdapter) recyclerView.getAdapter()).setList(views);
    }

    @Override
    public void showLoading() {
        loadingView.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideLoading() {
        loadingView.setVisibility(View.GONE);
    }

    @Override
    public void startGrocerySearch(int index) {
        startActivityForResult(GrocerySearchActivity.newIngredientReplaceIntent(this, GroceryEntity.GroceryEntityType.TYPE_COMBINED, index), Constants.ADD_REPLACE_INGREDIENT_INTENT_RESULT);
    }

    @Override
    public void startBarcodeScan(int index) {
        startActivityForResult(BarcodeScannerActivity.newReplaceIngredientSearchIntent(this, index), Constants.ADD_REPLACE_INGREDIENT_INTENT_RESULT);
    }

    @Override
    public void showSuccessfulSearchDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);

        builder.setMessage(getString(R.string.dialog_message_automated_ingredient_search_success));
        builder.setPositiveButton(getString(R.string.dialog_action_ok), new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                presenter.onSuccessfulSearchDialogOKClicked();
            }
        });
        builder.setCancelable(false);
        builder.create().show();
    }

    @Override
    public void setupViewsInRecyclerViewForAdaption(ArrayList<IngredientDisplayModel> ingredients, HashMap<Integer, Integer> failReasons) {
        ArrayList<IngredientSearchItemWrapper> views = new ArrayList<>();

        for(int i = 0; i < ingredients.size(); i++) {
            if (ingredients.get(i) instanceof ManualIngredientDisplayModel) {
                views.add(new IngredientSearchItemWrapper(IngredientSearchItemWrapper.ItemType.FAILED_INGREDIENT,
                        ingredients.get(i),
                        failReasons.get(i)));
            } else {
                views.add(new IngredientSearchItemWrapper(IngredientSearchItemWrapper.ItemType.FOUND_INGREDIENT, ingredients.get(i)));
            }
        }

        ((IngredientSearchListAdapter) recyclerView.getAdapter()).setList(views);
    }

    @Override
    public void showErrorWhileSavingRecipe() {
        Toast.makeText(this, getString(R.string.error_message_unknown_error), Toast.LENGTH_SHORT).show();
    }

    @Override
    public void finishActivity() {
        finish();
        setResult(Constants.CLOSE_SELF_RESULT, getIntent());
    }

    @Override
    public void showSaveButton() {
        saveButton.setVisibility(View.VISIBLE);
    }

    @Override
    public void showUnsuccessfulSearchDiaog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);

        builder.setMessage(getString(R.string.dialog_message_automated_ingredient_search_unsuccess));
        builder.setPositiveButton(getString(R.string.dialog_action_ok), new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {

            }
        });
        builder.setCancelable(false);
        builder.create().show();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home)
            this.finish();

        return super.onOptionsItemSelected(item);
    }

    @OnClick(R.id.save_button) public void onSaveButtonClicked(){
        presenter.onSaveButtonClicked();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (data != null) {
            switch (requestCode) {
                case Constants.ADD_REPLACE_INGREDIENT_INTENT_RESULT:
                    presenter.replaceIngredient(data.getIntExtra(EXTRA_INDEX, 0),
                            data.getIntExtra(EXTRA_GROCERY_ID, 0),
                            data.getFloatExtra(EXTRA_AMOUNT, 0.0f),
                            data.getIntExtra(EXTRA_UNIT_ID, 0));
                    break;
            }
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        presenter.finish();
        TrackerApplication.get(this).releaseAutomatedIngredientSearchComponent();
    }

}

File: app/src/main/java/de/karzek/diettracker/presentation/main/cookbook/recipeManipulation/automatedIngredientSearch/AutomatedIngredientSearchActivity.kt
package de.karzek.diettracker.presentation.main.cookbook.recipeManipulation.automatedIngredientSearch

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.LinearLayoutManager
import de.karzek.diettracker.R
import de.karzek.diettracker.data.cache.model.GroceryEntity
import de.karzek.diettracker.presentation.TrackerApplication
import de.karzek.diettracker.presentation.common.BaseActivity
import de.karzek.diettracker.presentation.main.cookbook.recipeManipulation.automatedIngredientSearch.adapter.IngredientSearchListAdapter
import de.karzek.diettracker.presentation.main.cookbook.recipeManipulation.automatedIngredientSearch.adapter.itemWrapper.IngredientSearchItemWrapper
import de.karzek.diettracker.presentation.model.IngredientDisplayModel
import de.karzek.diettracker.presentation.model.ManualIngredientDisplayModel
import de.karzek.diettracker.presentation.model.RecipeDisplayModel
import de.karzek.diettracker.presentation.search.grocery.GrocerySearchActivity
import de.karzek.diettracker.presentation.search.grocery.barcodeScanner.BarcodeScannerActivity
import de.karzek.diettracker.presentation.util.Constants
import de.karzek.diettracker.presentation.util.Constants.INVALID_ENTITY_ID
import kotlinx.android.synthetic.main.activity_automated_ingredient_search.*
import kotlinx.android.synthetic.main.snippet_semi_transparent_loading_view.*
import java.util.*
import javax.inject.Inject

class AutomatedIngredientSearchActivity : BaseActivity(), AutomatedIngredientSearchContract.View {

    @Inject
    lateinit var presenter: AutomatedIngredientSearchContract.Presenter

    override fun setupActivityComponents() {
        TrackerApplication.get(this).createAutomatedIngredientSearchComponent().inject(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_automated_ingredient_search)

        setupSupportActionBar()
        setupRecyclerView()
        disableSaveButton()

        presenter.setView(this)
        val recipeDisplayModel = intent.extras!!.getParcelable<RecipeDisplayModel>(EXTRA_RECIPE)
        presenter.setRecipe(recipeDisplayModel)
        if (recipeDisplayModel!!.id != INVALID_ENTITY_ID)
            presenter.startEdit()
        else
            presenter.start()

        save_button.setOnClickListener { presenter.onSaveButtonClicked() }
    }

    private fun disableSaveButton() {
        save_button.alpha = .5f
        save_button.isClickable = false
    }

    override fun enableSaveButton() {
        save_button.alpha = 1f
        save_button.isClickable = true
    }

    private fun setupSupportActionBar() {
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setDisplayShowHomeEnabled(true)
        supportActionBar!!.setHomeAsUpIndicator(resources.getDrawable(R.drawable.ic_back_arrow_white, null))
        supportActionBar!!.title = getString(R.string.automated_ingredient_search_title)
    }

    private fun setupRecyclerView() {
        recycler_view.setHasFixedSize(true)
        recycler_view.layoutManager = LinearLayoutManager(this)
        recycler_view.adapter = IngredientSearchListAdapter(presenter, presenter, presenter)
    }

    override fun setupViewsInRecyclerView(ingredients: ArrayList<IngredientDisplayModel>) {
        val views = ArrayList<IngredientSearchItemWrapper>()

        for (ingredient in ingredients) {
            if (ingredient is ManualIngredientDisplayModel) {
                views.add(IngredientSearchItemWrapper(IngredientSearchItemWrapper.ItemType.LOADING_INGREDIENT, ingredient))
            } else {
                views.add(IngredientSearchItemWrapper(IngredientSearchItemWrapper.ItemType.FOUND_INGREDIENT, ingredient))
            }
        }

        (recycler_view.adapter as IngredientSearchListAdapter).setList(views)
    }

    override fun showLoading() {
        loading_view.visibility = View.VISIBLE
    }

    override fun hideLoading() {
        loading_view.visibility = View.GONE
    }

    override fun startGrocerySearch(index: Int) {
        startActivityForResult(GrocerySearchActivity.newIngredientReplaceIntent(this, GroceryEntity.GroceryEntityType.TYPE_COMBINED, index), Constants.ADD_REPLACE_INGREDIENT_INTENT_RESULT)
    }

    override fun startBarcodeScan(index: Int) {
        startActivityForResult(BarcodeScannerActivity.newReplaceIngredientSearchIntent(this, index), Constants.ADD_REPLACE_INGREDIENT_INTENT_RESULT)
    }

    override fun showSuccessfulSearchDialog() {
        val builder = AlertDialog.Builder(this)

        builder.setMessage(getString(R.string.dialog_message_automated_ingredient_search_success))
        builder.setPositiveButton(getString(R.string.dialog_action_ok)) { _, _ -> presenter.onSuccessfulSearchDialogOKClicked() }
        builder.setCancelable(false)
        builder.create().show()
    }

    override fun setupViewsInRecyclerViewForAdaption(ingredients: ArrayList<IngredientDisplayModel>, failReasons: HashMap<Int, Int>) {
        val views = ArrayList<IngredientSearchItemWrapper>()

        for (i in ingredients.indices) {
            if (ingredients[i] is ManualIngredientDisplayModel) {
                views.add(IngredientSearchItemWrapper(IngredientSearchItemWrapper.ItemType.FAILED_INGREDIENT,
                        ingredients[i],
                        failReasons[i]!!))
            } else {
                views.add(IngredientSearchItemWrapper(IngredientSearchItemWrapper.ItemType.FOUND_INGREDIENT, ingredients[i]))
            }
        }

        (recycler_view.adapter as IngredientSearchListAdapter).setList(views)
    }

    override fun showErrorWhileSavingRecipe() {
        Toast.makeText(this, getString(R.string.error_message_unknown_error), Toast.LENGTH_SHORT).show()
    }

    override fun finishActivity() {
        finish()
        setResult(Constants.CLOSE_SELF_RESULT, intent)
    }

    override fun showSaveButton() {
        save_button.visibility = View.VISIBLE
    }

    override fun showUnsuccessfulSearchDiaog() {
        val builder = AlertDialog.Builder(this)

        builder.setMessage(getString(R.string.dialog_message_automated_ingredient_search_unsuccess))
        builder.setPositiveButton(getString(R.string.dialog_action_ok)) { _, _ -> }
        builder.setCancelable(false)
        builder.create().show()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home)
            this.finish()

        return super.onOptionsItemSelected(item)
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (data != null) {
            when (requestCode) {
                Constants.ADD_REPLACE_INGREDIENT_INTENT_RESULT -> presenter.replaceIngredient(data.getIntExtra(EXTRA_INDEX, 0),
                        data.getIntExtra(EXTRA_GROCERY_ID, 0),
                        data.getFloatExtra(EXTRA_AMOUNT, 0.0f),
                        data.getIntExtra(EXTRA_UNIT_ID, 0))
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        presenter.finish()
        TrackerApplication.get(this).releaseAutomatedIngredientSearchComponent()
    }

    companion object {

        const val EXTRA_INDEX = "EXTRA_INDEX"
        const val EXTRA_AMOUNT = "EXTRA_AMOUNT"
        const val EXTRA_UNIT_ID = "EXTRA_UNIT_ID"
        const val EXTRA_GROCERY_ID = "EXTRA_GROCERY_ID"
        const val EXTRA_RECIPE = "EXTRA_RECIPE"

        fun newIntent(context: Context, recipe: RecipeDisplayModel): Intent {
            val intent = Intent(context, AutomatedIngredientSearchActivity::class.java)
            intent.putExtra(EXTRA_RECIPE, recipe)

            return intent
        }
    }

}
